### PR TITLE
remove OnceInitializedOnceDisposed and instead import services via MEF

### PR DESCRIPTION
**Pull request by jmarolf**
**Customer scenario**

Customer notices that projects take a very long time to load.

**Bugs this fixes:** 

Fixes https://github.com/dotnet/project-system/issues/2421

**Workarounds, if any**

None

**Risk**

This is changing MEF imports so I would need manual verification of the scenario to ensure that nothing has regressed.

**Regression**

This is a perf regression in 15.3 because of new code in the dependencies node.

**Performance impact**

This will improve perf for project load and restore.

**Root cause analysis:**

Perviously we would switch to the UI thread, acquire the component catalog, and manually pick out the types we needed from the catalog while still on the UI thread.  Since we were in a pumping wait this would not deadlock the operation, but all threads waiting for this opertation would be blocked, causing the hill climbing algorithm in the threadpool to allocate more threads as no forward progress was being made.  Operations happening on the UI thread would also take longer as we where adding more work that needed to happen there.  This change ensures that the nuget types we need are composed by MEF when we are created so there in no wait.

**How was the bug found?**

dogfooding

### CODE REVIEW

**Comment by jmarolf** [comment]
@dotnet/project-system please review

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/CrossTarget/TargetFrameworkProvider.cs [hunk 1]
```diff
@@ -3,71 +3,32 @@
...
-using Microsoft.VisualStudio.ComponentModelHost;
-using Microsoft.VisualStudio.Shell;
 using NuGet.VisualStudio;
 
 namespace Microsoft.VisualStudio.ProjectSystem.VS.Tree.Dependencies.CrossTarget
 {
     [Export(typeof(ITargetFrameworkProvider))]
     [AppliesTo(ProjectCapability.DependenciesTree)]
-    internal class TargetFrameworkProvider : OnceInitializedOnceDisposed, ITargetFrameworkProvider
+    internal sealed class TargetFrameworkProvider : ITargetFrameworkProvider
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/CrossTarget/TargetFrameworkProvider.cs]: Why sealed?

**Comment by abpiskunov** [comment]
How do you ensure there is only one instance of ITargetFrameworkProvider? Also, is it safe to just import interop com objects from nuget? If they are not available by some resson, nuget version uninstaled, for example - VS will crash, thats why i had that logic there.

Thanks,
Anton

On Jun 12, 2017, at 5:47 PM, David Kean <notifications@github.com<mailto:notifications@github.com>> wrote:


@davkean approved this pull request.

**Comment by davkean** [comment]
There will only be one instance, parts are shared by default.

NuGet is now a required component, if it uninstalled, then we we're going to crash on startup when we nominate for restore.

**Comment by jmarolf** [comment]
@abpiskunov what is the case where MEF creates more than one `ITargetFrameworkProvider`? The only time that would happen was if multiple classes implemented `ITargetFrameworkProvider` and exported themselves.  As it stands today, this type will be created once when the host does the composition.

I don't know how the project system is supposed to do much of anything if nuget is unavailable.  The only sane thing would be to crash.  Otherwise we will just get null reference exceptions later that we will need to track down.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/CrossTarget/TargetFrameworkProvider.cs (hunk 1 above)
Comment (jmarolf) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/CrossTarget/TargetFrameworkProvider.cs]: This class used to have protected members on it but had no derived class.  I changed those members to private and saw no case where we need to derive from this class.  I can remove it :)

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/CrossTarget/TargetFrameworkProvider.cs [hunk 2]
```diff
@@ -3,71 +3,32 @@
...
-        protected override void Dispose(bool disposing)
-        {
-            if (disposing)
-            {
-                NugetComparer = null;
-                NugetFrameworkParser = null;
-            }
-        }
+        private readonly object _targetsLock = new object();
+        private List<ITargetFramework> CachedTargetFrameworks = new List<ITargetFramework>();
```
Comment (jmarolf) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/CrossTarget/TargetFrameworkProvider.cs]: 2 questions about `CachedTargetFrameworks`

1. I believe we are using the default string comparer below, don't we want to use `OrdinalIgnoreCase`?
```C#
targetFramework = CachedTargetFrameworks.FirstOrDefault(x => x.Equals(shortOrFullName));
```
2.  below adds a target framework to the cache if the `ITargetFramework` does not match our `shortOrFullName`.  Is there ever a case where we expect two different `ITargetFramework`s to resolve to the same string?  If not, couldn't this be a dictionary instead of a list?
```C#
var frameworkName = NugetFrameworkParser.ParseFrameworkName(shortOrFullName);
if (frameworkName != null)
{
    var shortName = NugetFrameworkParser.GetShortFrameworkName(frameworkName);
    targetFramework = new TargetFramework(frameworkName, shortName);
    // remember target framework - there can not bee too many of them across the solution.
    CachedTargetFrameworks.Add(targetFramework);
}
```

**Comment by abpiskunov** [comment]
There will be small number of unique target frameworks thats why i did not have dictionary originally.. but you can change that. I believe that shortnames are unique same as fullmonikers.

Thanks,
Anton

On Jun 12, 2017, at 6:13 PM, Jonathon Marolf <notifications@github.com<mailto:notifications@github.com>> wrote:


@jmarolf commented on this pull request.

________________________________

In src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/CrossTarget/TargetFrameworkProvider.cs<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fdotnet%2Fproject-system%2Fpull%2F2429%23discussion_r121563507&data=02%7C01%7Canton.piskunov%40microsoft.com%7Caa7a089eb53b4032daee08d4b1f95ccb%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636329131933303790&sdata=so%2ByVzlo1OmL%2BWAMvCh1I8Po4cy%2FWjIjbrP6jW7HeLk%3D&reserved=0>:

>
-        protected override void Dispose(bool disposing)
-        {
-            if (disposing)
-            {
-                NugetComparer = null;
-                NugetFrameworkParser = null;
-            }
-        }
+        private readonly object _targetsLock = new object();
+        private List<ITargetFramework> CachedTargetFrameworks = new List<ITargetFramework>();


2 questions about CachedTargetFrameworks

  1.  I believe we are using the default string comparer below, don't we want to use OrdinalIgnoreCase?

targetFramework = CachedTargetFrameworks.FirstOrDefault(x => x.Equals(shortOrFullName));

  1.  below adds a target framework to the cache if the ITargetFramework does not match our shortOrFullName. Is there ever a case where we expect two different ITargetFrameworks to resolve to the same string? If not, couldn't this be a dictionary instead of a list?

var frameworkName = NugetFrameworkParser.ParseFrameworkName(shortOrFullName);
if (frameworkName != null)
{
    var shortName = NugetFrameworkParser.GetShortFrameworkName(frameworkName);
    targetFramework = new TargetFramework(frameworkName, shortName);
    // remember target framework - there can not bee too many of them across the solution.
    CachedTargetFrameworks.Add(targetFramework);
}

**Comment by jmarolf** [comment]
nuget corruption error

```
error MSB4011: "C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\MSBuild\15.0\bin\Microsoft.Common.targets" cannot be imported again. It was already imported at "C:\Users\dotnet-bot\.nuget\packages\microbuild.plugins.swixbuild\1.0.147\build\Microsoft.VisualStudio.Setup.Tools.targets (113,3)". This is most likely a build authoring error. This subsequent import will be ignored. [D:\j\workspace\windows_debug---f7bc267c\src\VsixV3\EditorsPackage\Microsoft.VisualStudio.Editors.vsmanproj]
```

**Comment by jmarolf** [comment]
retest this please

**Comment by jmarolf** [comment]
@dotnet-bot retest this please

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/CrossTarget/TargetFrameworkProvider.cs [hunk 3]
```diff
@@ -76,22 +38,30 @@ public ITargetFramework GetTargetFramework(string shortOrFullName)
             ITargetFramework targetFramework = null;
             try
             {
+                // Fast path assuming that we've encountered this name before.
+                if (CachedTargetFrameworks.TryGetValue(shortOrFullName, out targetFramework))
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/CrossTarget/TargetFrameworkProvider.cs]: Dictionary is not thread safe for a reader and writer at the same time, I think you'd better off using a ConcurrentDictionary.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/CrossTarget/TargetFrameworkProvider.cs [hunk 4]
```diff
@@ -76,22 +38,30 @@ public ITargetFramework GetTargetFramework(string shortOrFullName)
...
+                    // returning.
                     var frameworkName = NugetFrameworkParser.ParseFrameworkName(shortOrFullName);
                     if (frameworkName != null)
                     {
                         var shortName = NugetFrameworkParser.GetShortFrameworkName(frameworkName);
                         targetFramework = new TargetFramework(frameworkName, shortName);
-                        // remember target framework - there can not bee too many of them accross the solution.
-                        CachedTargetFrameworks.Add(targetFramework);
+                        // Remember target framework, there can not be too many of them across the solution.
+                        CachedTargetFrameworks[shortOrFullName] = targetFramework;
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/CrossTarget/TargetFrameworkProvider.cs]: Why was this changed to a set instead of Add?

**Comment by davkean** [comment]
Can we separate the dictionary changes into a seperate PR for 15.6 - isn't not related to the original bug you are fixing.

**Comment by srivatsn** [comment]
I agree let's leave the dictionary change out of this.

**Comment by jmarolf** [comment]
dictionary change removed

**Comment by srivatsn** [comment]
Tagging @MattGertz for 15.3 - this is a perf regression.
